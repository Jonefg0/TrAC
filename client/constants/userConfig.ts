export const baseUserConfig = {
  SHOW_PROGRESS_STUDENT_CYCLE: false,
  SHOW_DROPOUT: false,
  SHOW_DOWNLOAD: false,
  SHOW_GROUPED_VIEW: false,
  SHOW_STUDENT_COMPLEMENTARY_INFORMATION: false,
  SHOW_GROUPED_COMPLEMENTARY_INFO: false,
  SHOW_STUDENT_LIST: false,
  FOREPLAN: false,
  FOREPLAN_ADVICES_COMPARE_BY_PERFORMANCE: false,
  FOREPLAN_COURSE_STATS: false,
  FOREPLAN_COURSE_FAIL_RATE_STATS: true,
  FOREPLAN_COURSE_EFFORT_STATS: true,
  FOREPLAN_SUMMARY_LIST: false,
  FOREPLAN_SUMMARY_BADGES: true,
  FOREPLAN_SUMMARY_WAFFLE_CHART: true,
  FOREPLAN_SUMMARY_ADVICE: true,
  FOREPLAN_FUTURE_COURSE_PLANIFICATION: true,
  FOREPLAN_PREDICT_CURRENT_COURSE: true,
  FEEDBACK: true,
};

export type UserConfig = typeof baseUserConfig &
  Record<string, boolean | string | number>;
