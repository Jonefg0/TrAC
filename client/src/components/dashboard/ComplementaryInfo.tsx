import { AnimatePresence, motion } from "framer-motion";
import React, { FC, memo, useContext, useEffect, useState } from "react";

import { Flex, Stack, Text } from "@chakra-ui/react";

import { ConfigContext } from "../../context/Config";
import { setTrackingData, track } from "../../context/Tracking";

export const ComplementaryInfo: FC<{
  educational_system?: string | null;
  institution?: string | null;
  months_to_first_job?: number | null;
  type_admission?: string | null;
}> = memo(
  ({
    educational_system,
    institution,
    months_to_first_job,
    type_admission,
  }) => {
    const {
      COMPLEMENTARY_INFORMATION,
      COMPLEMENTARY_INFORMATION_BACKGROUND_COLOR,
      COMPLEMENTARY_INFORMATION_EMPLOYED_EDUCATIONAL_SYSTEM,
      COMPLEMENTARY_INFORMATION_EMPLOYED_INSTITUTION,
      COMPLEMENTARY_INFORMATION_EMPLOYED_MONTHS,
      COMPLEMENTARY_INFORMATION_TEXT_COLOR,
      COMPLEMENTARY_INFORMATION_TYPE_ADMISSION,
    } = useContext(ConfigContext);

    const [show, setShow] = useState(false);

    useEffect(() => {
      setTrackingData({
        showingStudentComplementaryInformation: show,
      });
    }, [show]);

    return (
      <Flex alignItems="center" ml="1em">
        <Flex
          id="complementary_information"
          backgroundColor={COMPLEMENTARY_INFORMATION_BACKGROUND_COLOR}
          boxShadow={
            show
              ? "0px 0px 2px 1px rgb(174,174,174)"
              : "2px 3px 2px 1px rgb(174,174,174)"
          }
          borderRadius={show ? "5px 5px 5px 5px" : "0px 5px 5px 0px"}
          alignItems="center"
          onClick={() => {
            setShow((show) => !show);
            track({
              action: "click",
              effect: show
                ? "close-studentComplementaryInfo"
                : "open-studentComplementaryInfo",
              target: "studentComplementaryInfo",
            });
          }}
          color={COMPLEMENTARY_INFORMATION_TEXT_COLOR}
          cursor="pointer"
          transition="box-shadow 0.2s ease-in-out"
          data-testid="BoxContainer"
        >
          <Stack
            className="unselectable"
            isInline
            pt={10}
            pb={10}
            fontFamily="Lato"
          >
            <Text
              minWidth="90px"
              height="150px"
              m={0}
              ml={4}
              textAlign="center"
              fontWeight="bold"
              className="verticalText"
              fontSize="1.1em"
            >
              {COMPLEMENTARY_INFORMATION}
            </Text>
            <AnimatePresence>
              {show && (
                <motion.div
                  key="employed-text"
                  initial={{
                    opacity: 0,
                  }}
                  animate={{ opacity: 1 }}
                  exit={{
                    opacity: 0,
                  }}
                >
                  {type_admission && (
                    <Text width="290px" pl={5} pb={0} mb={0}>
                      {COMPLEMENTARY_INFORMATION_TYPE_ADMISSION}{" "}
                      {type_admission}
                    </Text>
                  )}

                  {educational_system && (
                    <Text width="290px" pl={5} pb={0} mb={0}>
                      {COMPLEMENTARY_INFORMATION_EMPLOYED_EDUCATIONAL_SYSTEM}{" "}
                      {educational_system}
                    </Text>
                  )}

                  {institution && (
                    <Text data-testid="inst" width="290px" pl={5} pb={0} mb={0}>
                      {COMPLEMENTARY_INFORMATION_EMPLOYED_INSTITUTION}{" "}
                      {institution}
                    </Text>
                  )}

                  {months_to_first_job && (
                    <Text width="290px" pl={5} pb={0} mb={0}>
                      {COMPLEMENTARY_INFORMATION_EMPLOYED_MONTHS}{" "}
                      {months_to_first_job}
                    </Text>
                  )}
                </motion.div>
              )}
            </AnimatePresence>
          </Stack>
        </Flex>
      </Flex>
    );
  }
);

export default ComplementaryInfo;
