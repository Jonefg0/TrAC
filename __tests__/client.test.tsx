import "@testing-library/jest-dom/extend-expect";

import React from "react";
import waitForExpect from "wait-for-expect";

import { MockedProvider } from "@apollo/react-testing";
import {
  act,
  cleanup,
  render,
  screen,
  fireEvent,
} from "@testing-library/react";

import { UserType } from "../client/constants";
import { baseConfig } from "../client/constants/baseConfig";
import { baseUserConfig } from "../client/constants/userConfig";
import {
  AllUsersAdminDocument,
  CheckUnlockDocument,
  CurrentUserDocument,
} from "../client/src/graphql";
import AdminPage from "../client/src/pages/admin";
import LoginPage from "../client/src/pages/login";
import UnlockPage from "../client/src/pages/unlock/[email]/[unlockKey]";
import ComplementaryInfo from "../client/src/components/dashboard/ComplementaryInfo";
Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

afterEach(async () => {
  await cleanup();
});

describe("unlock", () => {
  test("renders correctly", async () => {
    await act(async () => {
      const { getByText } = render(
        <MockedProvider
          mocks={[
            {
              request: {
                query: CurrentUserDocument,
              },
              result: {
                data: {
                  currentUser: {
                    user: null,
                  },
                },
              },
            },
            {
              request: {
                query: CheckUnlockDocument,
                variables: {
                  email: "asd@gmail.com",
                  unlockKey: "XwPp9xazJ0ku5CZnlmgAx2Dld8SHkAeT",
                },
              },
              result: {
                data: {
                  checkUnlockKey: null,
                },
              },
            },
          ]}
          addTypename={false}
        >
          <UnlockPage
            email="asd@gmail.com"
            unlockKey="XwPp9xazJ0ku5CZnlmgAx2Dld8SHkAeT"
          />
        </MockedProvider>
      );

      await waitForExpect(async () => {
        const NewPasswordFieldLabel = getByText(
          baseConfig.UNLOCK_NEW_PASSWORD_LABEL
        );

        expect(NewPasswordFieldLabel).toBeTruthy();
      });
    });
  });
});

describe("login", () => {
  test("renders correctly", async () => {
    await act(async () => {
      const { getByText } = render(
        <MockedProvider
          mocks={[
            {
              request: {
                query: CurrentUserDocument,
              },
              result: {
                data: {
                  currentUser: {
                    user: null,
                  },
                },
              },
            },
          ]}
          addTypename={false}
        >
          <LoginPage />
        </MockedProvider>
      );

      await waitForExpect(async () => {
        const LoginButton = getByText(baseConfig.LOGIN_BUTTON);

        expect(LoginButton).toBeTruthy();
        expect(LoginButton).toHaveAttribute("disabled");
      });
    });
  });
});

describe("admin", () => {
  test("renders correctly", async () => {
    await act(async () => {
      const { getByText } = render(
        <MockedProvider
          mocks={[
            {
              request: {
                query: CurrentUserDocument,
              },
              result: {
                data: {
                  currentUser: {
                    user: {
                      email: "asd@gmail.com",
                      name: "name",
                      admin: true,
                      type: UserType.Director,
                      config: baseUserConfig,
                      student_id: "",
                      __typename: "User",
                    },
                    __typename: "AuthResult",
                  },
                },
              },
            },
            {
              request: {
                query: AllUsersAdminDocument,
              },
              result: {
                data: {
                  users: [],
                },
              },
            },
          ]}
          addTypename={true}
        >
          <AdminPage />
        </MockedProvider>
      );

      await waitForExpect(async () => {
        const UsersMenu = getByText("Users");

        expect(UsersMenu).toBeTruthy();
        expect(UsersMenu).toHaveClass("active");
      });
    });
  });
});

describe("Test <AdmissionDropout />", () => {
  test("Test Snapshot", () => {
    const tree = render(
      <ComplementaryInfo
        type_admission="PSU"
        educational_system="Publico"
        institution="Escuela"
        months_to_first_job={6}
      />
    );
    fireEvent.click(screen.getByTestId("BoxContainer"));
    expect(tree).toMatchSnapshot();
  });
});

test("test props en <AdmissionDropout />", () => {
  const type_admission = "PSU";
  const educational_system = "Publico";
  const institution = "Escuela";
  const months_to_first_job = 6;

  const { getByText, getByTestId } = render(
    <ComplementaryInfo
      type_admission={type_admission}
      educational_system={educational_system}
      institution={institution}
      months_to_first_job={months_to_first_job}
    />
  );
  fireEvent.click(getByTestId("BoxContainer"));
  expect(getByText("Tipo de ingreso: PSU".trim())).toBeInTheDocument();
  expect(getByText("Sistema educacional: Publico".trim())).toBeInTheDocument();
});

test("test null props complementary component", () => {
  const type_admission = null;
  const educational_system = "Publico";
  const institution = null;
  const months_to_first_job = null;

  const { getByText, getByTestId, queryAllByText } = render(
    <ComplementaryInfo
      type_admission={type_admission}
      educational_system={educational_system}
      institution={institution}
      months_to_first_job={months_to_first_job}
    />
  );

  fireEvent.click(getByTestId("BoxContainer"));
  const inst_value = queryAllByText("Tipo de ingreso:");
  expect(inst_value).toHaveLength(0);
  const education_system_value = queryAllByText(
    "Sistema educacional: Publico".trim()
  );
  expect(education_system_value).toHaveLength(1);
});

export {};
