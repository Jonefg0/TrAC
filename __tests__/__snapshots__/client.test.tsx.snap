// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <AdmissionDropout /> Test Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div
      id="chakra-toast-portal"
    >
      <ul
        id="chakra-toast-manager-top"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto; top: 0px; right: 0px; left: 0px;"
      />
      <ul
        id="chakra-toast-manager-top-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; top: 0px; left: 0px;"
      />
      <ul
        id="chakra-toast-manager-top-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; top: 0px; right: 0px;"
      />
      <ul
        id="chakra-toast-manager-bottom-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; bottom: 0px; left: 0px;"
      />
      <ul
        id="chakra-toast-manager-bottom"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto; bottom: 0px; right: 0px; left: 0px;"
      />
      <ul
        id="chakra-toast-manager-bottom-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; bottom: 0px; right: 0px;"
      />
    </div>
    <div>
      <div
        class="css-xtzu7m"
      >
        <div
          class="css-qfh9pt"
          data-testid="BoxContainer"
        >
          <div
            class="chakra-stack unselectable css-iquqh5"
          >
            <p
              class="chakra-text verticalText css-1mwd8m3"
            >
              Información Complementaria
            </p>
            <div
              style="opacity: 0;"
            >
              <p
                class="chakra-text css-1yyofq7"
              >
                Tipo de ingreso: 
                 
                PSU
              </p>
              <p
                class="chakra-text css-1yyofq7"
              >
                Evaluación inicial diagnóstica: 
                 
                10
              </p>
              <p
                class="chakra-text css-1yyofq7"
              >
                Evaluación nacional diagnóstica: 
                 
                20
              </p>
              <p
                class="chakra-text css-tcsrcc"
              >
                Sistema educacional:
                 
                Publico
              </p>
              <p
                class="chakra-text css-13kiv3l"
                data-testid="inst"
              >
                Nombre de institución:
                 
                Escuela
              </p>
              <p
                class="chakra-text css-1yyofq7"
              >
                Meses hasta encontrar trabajo:
                 
                6
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="css-xtzu7m"
    >
      <div
        class="css-qfh9pt"
        data-testid="BoxContainer"
      >
        <div
          class="chakra-stack unselectable css-iquqh5"
        >
          <p
            class="chakra-text verticalText css-1mwd8m3"
          >
            Información Complementaria
          </p>
          <div
            style="opacity: 0;"
          >
            <p
              class="chakra-text css-1yyofq7"
            >
              Tipo de ingreso: 
               
              PSU
            </p>
            <p
              class="chakra-text css-1yyofq7"
            >
              Evaluación inicial diagnóstica: 
               
              10
            </p>
            <p
              class="chakra-text css-1yyofq7"
            >
              Evaluación nacional diagnóstica: 
               
              20
            </p>
            <p
              class="chakra-text css-tcsrcc"
            >
              Sistema educacional:
               
              Publico
            </p>
            <p
              class="chakra-text css-13kiv3l"
              data-testid="inst"
            >
              Nombre de institución:
               
              Escuela
            </p>
            <p
              class="chakra-text css-1yyofq7"
            >
              Meses hasta encontrar trabajo:
               
              6
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
